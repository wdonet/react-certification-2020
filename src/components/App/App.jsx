import React from 'react';
import NavBar from '../NavBar';
import HomePage from '../../pages/Home';

function App() {
  return (
    <div>
      <NavBar />
      <HomePage />
    </div>
  );
}

export default App;
