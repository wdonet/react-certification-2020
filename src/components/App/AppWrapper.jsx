import React, { useReducer } from 'react';
import { appReducer } from '../../reducers/appReducer';
import { lightTheme, darkTheme } from '../../providers/themes';
import youtubeSearchService from '../../services/youtubeSearchService';
import AppContext from '../../providers/AppContext';
import { 
  SET_IS_FIRST_LOAD, 
  SET_SESSION, 
  SET_VIDEOS_LIST, 
  SWITCH_THEME 
} from '../../reducers/actionTypes';

const USER_SESSION_KEY = "user_session";
const initialState = {
  videosList: [],
  currentVideoId: undefined,
  isLightTheme: true,
  isFirstLoad: true,
  userSession: JSON.parse(window.sessionStorage.getItem(USER_SESSION_KEY)),
};

function AppWrapper(props) {
  const [{ 
    videosList, 
    isLightTheme, 
    isFirstLoad,
    userSession}, 
    dispatch] = useReducer(appReducer, initialState);

  const getThemeConfig = () => ({
    theme: isLightTheme ? lightTheme : darkTheme,
    switchTheme: () => dispatch({type: SWITCH_THEME}),
  });

  const search = async (query) => {
    const serviceResponse = await youtubeSearchService.search(query);
    serviceResponse.error 
    ? dispatch({ type: SET_VIDEOS_LIST, payload: [] })
    : dispatch({ type: SET_VIDEOS_LIST, payload: serviceResponse });
  }

  const firstSearch = async () => {
    let counter = 0;
    /* global gapi */
    /* eslint no-undef: "error" */
    while((!gapi || !gapi.client) && counter < 10){
      counter+=1;
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    ((!gapi || !gapi.client) || counter >= 10) 
    ? dispatch({ type: SET_VIDEOS_LIST, payload: [] })
    : search()
  }

  if(isFirstLoad){
    dispatch({ type: SET_IS_FIRST_LOAD, payload: false });
    firstSearch();
  }

  return (
    <AppContext.Provider value={
      { ...getThemeConfig(), 
        search,
        videosList,
        userSession,
        setUserSession: (data) => dispatch({type: SET_SESSION, payload: data}),
      }
    }>
      { props.children }
    </AppContext.Provider>
  );
}

export default AppWrapper;
