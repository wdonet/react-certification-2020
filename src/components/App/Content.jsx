import React from 'react';
import { Redirect, Route, Switch } from 'react-router';
import HomePlayer from '../HomeVideos/HomePlayer';
import HomeVideos from '../HomeVideos/HomeVideos';

const Content = () => {
    return (<div>
        <Switch>
            <Route exact path="/">
                <Redirect to="/home"/>
            </Route>
            <Route path="/home">
                <HomeVideos />
            </Route>
            <Route path="/player">
                <HomePlayer />
            </Route>
        </Switch>
    </div>);
}

export default Content;