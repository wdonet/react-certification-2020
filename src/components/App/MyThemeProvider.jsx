import React from 'react';
import { ThemeProvider } from 'styled-components';
import { useYouTube } from '../YouTube/YouTubeProvider';

const theme = {
  dark: {
    primary: '#4396f3',
    secondary: '#4a4646',
    color: '#f7f7f7',
  },
};

function MyThemeProvider({ children }) {
  const { state } = useYouTube();
  const { theme: stateTheme } = state;
  return <ThemeProvider theme={theme[stateTheme] || {}}>{children}</ThemeProvider>;
}

export default MyThemeProvider;
