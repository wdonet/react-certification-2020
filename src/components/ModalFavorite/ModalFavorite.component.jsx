import React from 'react';
import { Link } from 'react-router-dom';
import Styled from './ModalFavoriteStyled';

function ModalFavorite({ isLogin }) {
  return (
    <Styled.ModalFavorite>
      <Link to="/">
        <Styled.MenuContent>
          <Styled.SpanMenu>Home</Styled.SpanMenu>
        </Styled.MenuContent>
      </Link>
      {isLogin ? (
        <Link to="/favoriteVideo">
          <Styled.MenuContent>
            <Styled.SpanMenu>Favorite</Styled.SpanMenu>
          </Styled.MenuContent>
        </Link>
      ) : null}
    </Styled.ModalFavorite>
  );
}

export default ModalFavorite;
