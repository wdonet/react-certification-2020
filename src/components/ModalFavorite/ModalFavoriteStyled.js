import styled from 'styled-components';

const ModalFavorite = styled.div`
  width: 30%;
  height: 100%;
  background-color: white;
  left: 0;
  top: 0px;
  display: block;
  position: fixed;
  z-index: 1;
  overflow: auto;
`;

const MenuContent = styled.div`
  padding: 10px;
  height: 50px;
  border-bottom-style: solid;
  border-bottom-color: #e6e6e6;
`;

const SpanMenu = styled.span`
  color: black;
  font-size: 1rem;
  font-weight: 400;
`;

const Styled = { ModalFavorite, SpanMenu, MenuContent };
export default Styled;
