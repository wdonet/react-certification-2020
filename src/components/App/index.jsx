import React, { useEffect, useState }from 'react';
import HeaderBar from '../HeaderBar';
import Content from '../Content';
import { items } from '../../assets/mockdata/mockdata.json';
import VideoCard from '../VideoCard';
import filterByYear from '../../utils/filter';

function App() {
  const [recent, setRecent] = useState([]);
  const [inDetail, setInDetail] = useState(false);
  const [detailVideoId, setDetailVideoId] = useState("");
  const [detailTitle, setDetailTitle] = useState("");
  const [detailDescription, setDetailDescription] = useState("");
  
  
  useEffect(() => { 
      console.log('useEffect');
      fetch(`https://www.googleapis.com/youtube/v3/search?key=${process.env.REACT_APP_YOUTUBE_API_KEY2}&q=wizeline&part=snippet&maxResults=50&order=date&type=video`)
      .then((res) => res.json())
      .then((res) => {
        if (res.items) {
          setRecent(res.items);
        } else {
          setRecent(items);
        }
        
      }
    ).catch((emAll) => {});
  }, []);
  
  const UpdateQuery = (text) => {
    console.log(`New query: ${text}`);
    fetch(`https://www.googleapis.com/youtube/v3/search?key=${process.env.REACT_APP_YOUTUBE_API_KEY2}&q=${text}&part=snippet&maxResults=50&order=date&type=video`)
      .then((res) => res.json())
      .then((res) => {
        if (res.items) {
          setRecent(res.items);
        } else {
          setRecent(items);
        }
        setInDetail(false);
      });
  };

  const changeDetail = (indetail=false, videoId, title, description) => {
    // setInDetail(inDetail ? false : true);
    console.log(`title: ${title}`);
    console.log(`description: ${description}`);
    setInDetail(indetail);
    setDetailVideoId(videoId);
    setDetailTitle(title);
    setDetailDescription(description);
  };


  return (
    <React.StrictMode>
      <HeaderBar updateQuery={ UpdateQuery } />
      <Content inDetail={inDetail} detailVideoId={detailVideoId} gotodetail={changeDetail} detailTitle={detailTitle} detailDescription={ detailDescription}>{recent ?
        recent.map((video) => {
          return (
            <VideoCard
              key={video.etag}
              videoId ={video.id.videoId}
              title={video.snippet.title}
              description={video.snippet.description}
              image={video.snippet.thumbnails.high.url}
              gotodetail={ changeDetail }
            />
          )
        })
      :null}
      </Content>
    </React.StrictMode>
  );
}

export default App;
