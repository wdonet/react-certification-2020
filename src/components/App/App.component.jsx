import React, {
  useLayoutEffect,
  useState,
  useEffect,
  useCallback,
  createContext,
  useReducer,
} from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { createGlobalStyle, ThemeProvider } from 'styled-components';
import AuthProvider from '../../providers/Auth';
import HomePage from '../../pages/Home';
// import LoginPage from '../../pages/Login';
import NotFound from '../../pages/NotFound';
import SecretPage from '../../pages/Secret';
import VideoDetailViewPage from '../../pages/VideoDetailView/VideoDetailView.page';
import FavoriteVideo from '../../pages/FavoriteVideos/FavoriteVideo.page';
import FavoriteVideoDetail from '../../pages/FavoriteVideos/FavoriteVideoDetail.page';
import Private from '../Private';
import Fortune from '../Fortune';
import Header from '../Header';
import Layout from '../Layout';
import { random } from '../../utils/fns';
import useCallAPI from '../../utils/hooks/useCallApi';
import { initialState, reducer } from '../../reducers/reducerDarkLight';
// eslint-disable-next-line import/no-cycle
import Login from '../ModalLogin/ModalLogin.component.jsx';
import ModalFavorite from '../ModalFavorite/ModalFavorite.component';

export const AppContext = createContext();

const GlobalStyles = createGlobalStyle`
  html,body {
    font-family: Arial;
    text-align: center;
    margin: 0 0 15px 0;
    background: ${(props) => props.theme.backgroundColor};
    color: ${(props) => props.theme.textColor};
  }
`;

function App() {
  const [state, dispatch] = useReducer(reducer, initialState);
  const { currentTheme } = state;

  const [videos, setVideos] = useState([]);
  const gapi = useCallAPI();

  const getVideos = useCallback(
    async (part, maxResults, searchItem) => {
      const responseApi = await gapi.client.youtube.search.list({
        maxResults,
        part,
        q: searchItem,
      });
      setVideos(responseApi.result.items);
    },
    [gapi]
  );

  useEffect(() => {
    const part = ['id', 'snippet'];
    const maxResults = 29;
    const searchItem = 'wizeline';

    if (gapi) getVideos(part, maxResults, searchItem);
  }, [gapi, getVideos]);

  useLayoutEffect(() => {
    const { body } = document;

    function rotateBackground() {
      const xPercent = random(100);
      const yPercent = random(100);
      body.style.setProperty('--bg-position', `${xPercent}% ${yPercent}%`);
    }

    const intervalId = setInterval(rotateBackground, 3000);
    body.addEventListener('click', rotateBackground);

    return () => {
      clearInterval(intervalId);
      body.removeEventListener('click', rotateBackground);
    };
  }, []);

  return (
    <BrowserRouter>
      <AuthProvider>
        <Layout>
          <Switch>
            <Route exact path="/">
              <ThemeProvider theme={currentTheme}>
                <AppContext.Provider value={{ ...state, dispatch }}>
                  <Header getVideoCallBack={getVideos} />
                  <GlobalStyles />
                  <HomePage homeVideos={videos} />
                </AppContext.Provider>
              </ThemeProvider>
            </Route>
            <Route exact path="/video/:getVideoId/:title">
              <ThemeProvider theme={currentTheme}>
                <AppContext.Provider value={{ ...state, dispatch }}>
                  <Header getVideoCallBack={getVideos} />
                  <GlobalStyles />
                  <VideoDetailViewPage
                    detailVideo={videos}
                    isLogin={state.currentSession}
                  />
                </AppContext.Provider>
              </ThemeProvider>
            </Route>
            <Route exact path="/loginPage">
              <AppContext.Provider value={{ ...state, dispatch }}>
                <Login isLogin={state.currentSession} />
                {/* <LoginPage /> */}
              </AppContext.Provider>
            </Route>
            <Route exact path="/favoriteVideo">
              <ThemeProvider theme={currentTheme}>
                <AppContext.Provider value={{ ...state, dispatch }}>
                  <Header getVideoCallBack={getVideos} />
                  <FavoriteVideo />
                  <GlobalStyles />
                </AppContext.Provider>
              </ThemeProvider>
            </Route>
            <Route
              exact
              path="/favoriteVideoDetail/:linkFavoriteVideoId/:linkFavoiteVideoTitle"
            >
              <ThemeProvider theme={currentTheme}>
                <AppContext.Provider value={{ ...state, dispatch }}>
                  <Header getVideoCallBack={getVideos} />
                  <FavoriteVideoDetail />
                  <GlobalStyles />
                </AppContext.Provider>
              </ThemeProvider>
            </Route>
            <Route exact path="/favoriteModal">
              <ThemeProvider theme={currentTheme}>
                <AppContext.Provider value={{ ...state, dispatch }}>
                  <ModalFavorite isLogin={state.currentSession} />
                  <Header getVideoCallBack={getVideos} />
                  <GlobalStyles />
                </AppContext.Provider>
              </ThemeProvider>
            </Route>
            {/* <Route exact path="/login">
              <LoginPage />
            </Route> */}
            <Private exact path="/secret">
              <SecretPage />
            </Private>
            <Route path="*">
              <NotFound />
            </Route>
          </Switch>
          <Fortune />
        </Layout>
      </AuthProvider>
    </BrowserRouter>
  );
}

export default App;
