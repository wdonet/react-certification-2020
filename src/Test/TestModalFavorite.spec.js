import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import ModalFavorite from '../components/ModalFavorite/ModalFavorite.component';
import { AppContext } from '../components/App/App.component';

describe('Modal Favorite Components', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(
      <Router>
        <AppContext.Provider value={false}>
          <ModalFavorite />
        </AppContext.Provider>
      </Router>
    );
    expect(asFragment(<ModalFavorite />)).toMatchSnapshot();
  });

  it('Home Link is present', () => {
    render(
      <Router>
        <AppContext.Provider value={false}>
          <ModalFavorite />
        </AppContext.Provider>
      </Router>
    );
    expect(screen.getAllByRole('link')).toBeTruthy();
  });
});
