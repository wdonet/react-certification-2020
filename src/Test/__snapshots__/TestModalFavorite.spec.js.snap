// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Favorite Components should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ gTUOmw"
  >
    <a
      href="/"
    >
      <div
        class="sc-gsTCUz hlMnr"
      >
        <span
          class="sc-dlfnbm bwuiqW"
        >
          Home
        </span>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
